learning-from-errors/overlapping-instances/article.md

```diff
@@ -0,0 +1,626 @@
+be overlapping. And since this does not match, we cannot eliminate `Printable a`.
+
+The next one is `Printable (f a)`, and we cannot eliminate it since the other
+instance `a` is not a substitution instance of `f a`. If something is expecting `f a`
+you cannot give `a` to it. Or in other words, `a` is not more specific than `f a`, but
+instead, it is more general.
+
+So after this rule, both instances remain, and the next rule says:
+
+```
+If all the remaining candidates are incoherent, the search succeeds, returning an arbitrary surviving candidate.
+```
+
+None of our instances are marked `Incoherent`, so we proceed to the next rule.
+
+```
+If more than one non-incoherent candidate remains, the search fails.
+```
+
+Considering that we have two such instances now, the lookup fails here.
+
+Let us try adding an `OVERLAPPING` pragma to the instance for `f a`.
+
+```
+instance {-# OVERLAPPING #-} Functor f => Printable (f a) where
+  printMe _ = putStrLn "Instance for a Functor"
+```
+
+And now we get the error:
+
+```
+• No instance for (Functor MyType) arising from a use of ‘printMe’
+• In the expression: printMe (MyType 'c')
+  In an equation for ‘main’: main = printMe (MyType 'c')
+```
+
+We can see that adding `OVERLAPPING` pragma enabled the elimination of the
+instance for `Printable a` at step 3. But the remaining instance `Functor f => Printable (f a)`
+failed to work, but this failure happens at a later phase: when constraints are matched and
+after GHC has picked an instance. This is why we get a `No instance for Functor MyType`
+error instead of an overlapping instance error.
+
+Here we come across a crucial aspect of instance resolution: the algorithm
+never backtracks. When the algorithm failed during matching the constraints, if it could
+backtrack, it could have picked the instance for `Printable a`, which was
+eliminated at a later stage, in favor of the failed instance for `f a`. But instead, the
+algorithm just fails.
+
+### How to fix it?
+
+We can either remove the instance for `f a`, which makes the algorithm pick the
+instance for `Printable a`. Or else we can add a Functor instance for `MyType a`
+if it makes sense.
+
+## Higher-kinded overlapping instances
+
+Don't worry if you have never heard of that name. That's because I just made it up.
+
+To demonstrate this, we unfortunately need a bit more elaborate setup, and
+frankly, this example is a bit contrived. Anyway, so we have this code below
+which throws our beloved error:
+
+```hs
+{-# LANGUAGE DataKinds              #-}
+{-# LANGUAGE FlexibleInstances      #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE KindSignatures         #-}
+{-# LANGUAGE MultiParamTypeClasses  #-}
+{-# LANGUAGE PolyKinds              #-}
+{-# LANGUAGE RankNTypes             #-}
+{-# LANGUAGE UndecidableInstances   #-}
+
+module Main (main) where
+
+import Data.Proxy
+import GHC.TypeLits
+
+class Printable n a where
```

Comment:
I think you’re introducing too many things at once. It would be more helpful if you first considered the interaction between overlapping instances and multi-param type classes, and only after that throw `PolyKinds` into the mix.

Also, using a contrived example goes against the spirit of this article, I believe. The first part introduced a simple class `Printable`, the intention behind which was pretty clear, and used this class to explain peculiar GHC behavior. But now you’re just throwing some code together without purpose, so why would I care what GHC has to say about it?

I’m sure it’s possible to find a way to extend `Printable` in a meaningful way. For example, your phantom parameter passed by `Proxy` could be printing configuration

```
class Printable n a where
  printMe :: Proxy n -> a -> IO ()

instance Printable 10 Int where  -- print in base 10
  printMe _ x = print x

instance Printable 16 Int where  -- print in base 16
  printMe _ x = putStrLn (showHex x “”)
```

That’s just an example. But surely there are ways to make the example compelling.